src/Buckle/Buckle/CodeAnalysis/Binding/Binder.cs

```diff
@@ -860,18 +856,48 @@ private struct BinderState {
 
     private BoundStatement BindDoWhileStatement(DoWhileStatement statement, bool insideInline = false) {
         var body = BindLoopBody(statement.body, out var breakLabel, out var continueLabel, insideInline);
-        var conditionValue = RemoveNullability(statement.condition);
-        var condition = BindCast(conditionValue, BoundTypeClause.NullableBool);
+        var condition = BindDoubleCastRemovingNullability(
+            statement.condition, BoundTypeClause.NullableBool, BoundTypeClause.Bool);
 
         return new BoundDoWhileStatement(body, condition, breakLabel, continueLabel);
     }
 
+    private BoundExpression BindDoubleCastRemovingNullability(
```

Comment:
This function basically goes from
nullable type -> [NotNull]boolean
in 2 steps so that the errors show a nullable boolean in the message, but under the hood it is a non-nullable boolean
the context doesn't really matter here as the function only does the casting part and is isolated